File: plugins/rummy-plugin/srummy/edu.wpi.always.srummy/edu/wpi/always/srummy/game/LayoffMove.java
package edu.wpi.always.srummy.game;

/**
 * A class for Rummy game layoff move.
 * @author Morteza Behrooz
 * @version 1.0
 */

public class LayoffMove extends Move{
	
	private final Card playedCard;
	private final Meld relatedMeld;
	
	
	public Card getCard(){
		return playedCard;
	}

	public Meld getItsMeld(){
		return relatedMeld;
	}
	
	public LayoffMove(Player thePlayer, Card aCard, Meld aMeld){
		
		super(thePlayer);
		this.playedCard = aCard;
		this.relatedMeld = aMeld;

	}
	
	/**
	 * Checks if another LayoffMove is semantically 
	 * the same one by Card, Player and Meld criteria.
	 * @param Object someMove
	 * @return true if same Card, Player and Meld (semantically same Layoff).
	 * @see TestLayOffMove
	 */
	@Override
	public boolean equals(Object someMove){
		
		if(someMove == null 
				|| !(someMove instanceof LayoffMove))
			return false;
		
		LayoffMove supposedlySameLayOffMove 
				= (LayoffMove) someMove;
		
		if(!(supposedlySameLayOffMove.getPlayer()
				== this.player))
			return false;
		
		if(!supposedlySameLayOffMove.getCard()
				.equals(this.playedCard))
			return false;
		
		if(!supposedlySameLayOffMove.getItsMeld()
				.equals(this.playedCard))
			return false;
		
		return true;
	
	}
	
	/**
	 * Overides hashCode() of Object as equals() is overriden
	 * (In Java, hashCode() must return the same 
	 * value for two objects which are equal)
	 * @return int hashCode
	 * @see TestLayoffMove
	 */
	@Override
	public int hashCode(){
		
		return player.ordinal()
				+ 2 * playedCard.hashCode()
				+ 2 * relatedMeld.hashCode()
				;
		
	}
	
	/**
	 * States the layoff move as String. 
	 * @return String of layoff statement
	 */
	@Override
	public String toString(){
		
		return 
				this.player.name()
				+ " doing a layoff with adding "
				+ playedCard.toString()
				+ " to the "
				+ relatedMeld.toString()
		;
		
	}

	@Override
	public void happen(GameState gameState) {
		
		gameState.layOff(player, playedCard, relatedMeld);
		
	}
	

}

File: plugins/rummy-plugin/Rummy.Game/LayOffMove.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Rummy
{
    public class LayOffMove : Move
    {
        private readonly Card _card;

        public Card Card
        {
            get { return _card; }
        }

        private readonly Meld _meld;

        public Meld Meld
        {
            get { return _meld; }
        }

        public Card GetCard()
        {
            return _card;
        }

        public LayOffMove(Player player, Card card, Meld meld)
            : base(player)
        {
            _card = card;
            _meld = meld;
        }

        public override bool Equals(Object obj)
        {
            if (obj == null || obj.GetType() != typeof(LayOffMove))
                return false;

            var theOther = (LayOffMove)obj;

            if (theOther.Player != this.Player)
                return false;

            if (!theOther._card.Equals(this._card))
                return false;

            if (!theOther._meld.Equals(this._meld))
                return false;

            return true;
        }

        public override int GetHashCode()
        {
            return (int)Player + _card.GetHashCode() * 2 + _meld.GetHashCode() * 2;
        }

        public override void Realize(GameState gameState)
        {
            gameState.LayOff(Player, _card, _meld);
        }

        public override string ToString()
        {
            return string.Format("Player {0} -- LayOff {1} to meld", Player, _card);
        }
    }
}
